### PR TITLE
[SPARK-48778][SQL][TESTS] Improve collation support testing - add unit tests for string expressions

### PR DESCRIPTION
**Pull request by uros-db**
### What changes were proposed in this pull request?
Add unit tests at Spark expression level.

### Why are the changes needed?
Enhance collation testing for both code gen and interpreted code path.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
New tests in `CollationStringExpressionsSuite.scala`

### Was this patch authored or co-authored using generative AI tooling?
Yes.

### CODE REVIEW

**Review by uros-db — COMMENTED** [review]
@cloud-fan ready for review

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java [hunk 1]
```diff
@@ -477,7 +477,7 @@ public static String genCode(final String string, final String delimiter,
       } else if (collation.supportsLowercaseEquality) {
         return String.format(expr + "Lowercase(%s, %s, %s)", string, delimiter, count);
       } else {
-        return String.format(expr + "ICU(%s, %s, %d, %s)", string, delimiter, count, collationId);
+        return String.format(expr + "ICU(%s, %s, %s, %d)", string, delimiter, count, collationId);
```
[the remark this answers is not in the document]
Comment (uros-db) [reply common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java:480]: sorry for the confusion, these changes are already merged in latest master as part of a separate PR

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java [hunk 2]
```diff
@@ -490,8 +490,7 @@ public static UTF8String execLowercase(final UTF8String string, final UTF8String
     }
     public static UTF8String execICU(final UTF8String string, final UTF8String delimiter,
         final int count, final int collationId) {
-      return CollationAwareUTF8String.subStringIndex(string, delimiter, count,
-        collationId);
+      return CollationAwareUTF8String.subStringIndex(string, delimiter, count, collationId);
```
[the remark this answers is not in the document]
Comment (uros-db) [reply common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java:493]: sorry for the confusion, these changes are already merged in latest master as part of a separate PR

File: common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (MaxGekk) [reply common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationSupport.java:480]: Could you rebase this PR on the recent master than, please.

File: sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala [hunk 3]
```diff
@@ -17,135 +17,130 @@
...
 import org.apache.spark.sql.test.SharedSparkSession
-import org.apache.spark.sql.types.{ArrayType, BinaryType, BooleanType, DataType, IntegerType, StringType}
+import org.apache.spark.sql.types._
 
 // scalastyle:off nonascii
 class CollationStringExpressionsSuite
-  extends QueryTest
-  with SharedSparkSession
-  with ExpressionEvalHelper {
+    extends QueryTest
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala:28]: Why did you add the indents. The style guide says about 2 spaces, see https://github.com/databricks/scala-style-guide?tab=readme-ov-file#spacing-and-indentation. Does scala style check complain about it?

File: sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala [hunk 4]
```diff
@@ -17,135 +17,130 @@
...
-      assert(sql(query).schema.fields.head.dataType.sameType(StringType(t.c)))
-      // Implicit casting
-      val arr = t.a.map(s => s"'$s'").mkString(", ")
-      query = s"SELECT concat_ws(collate('${t.s}', '${t.c}'), $arr)"
-      checkAnswer(sql(query), Row(t.result))
-      assert(sql(query).schema.fields.head.dataType.sameType(StringType(t.c)))
-      query = s"SELECT concat_ws('${t.s}', $arrCollated)"
-      checkAnswer(sql(query), Row(t.result))
-      assert(sql(query).schema.fields.head.dataType.sameType(StringType(t.c)))
+      // Unit test.
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala:45]: How about to split the expressions and end-to-end tests to 2 different suites like `StringExpressionsSuite` and `StringFunctionsSuite` (could be `CollationStringE2ESuite`). The expression test suite might cover more cases but e2e test suite should check just some cases.

File: sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala (hunk 3 above)
Comment (uros-db) [reply sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala]: fixed

**Comment by uros-db** [comment]
@cloud-fan, @MaxGekk and I have discussed this offline, so here's a quick rundown:

over the past few months, we've had many untested and mishandled cases for collation support in various expressions - so to avoid things like this, I propose that we should stay on the conservative side of testing with collations (at least for now)

given that this suite takes ~10s total to run on my machine locally, I think this is not too bad and that benefits outweight the extra computation load - to recap, `CollationStringExpressionsSuite.scala` currently handles two types of tests:
- Spark expression-level unit tests, that are used to test both code paths (interpreted and genCode)
- e2e SQL tests with literals, that are used to conveniently verify only the basic test case results

the other stuff, such as:
- e2e SQL tests with explicit `collate` expressions
- e2e SQL tests that verify collation type coercion & collation mismatches
- e2e SQL tests with table columns instead of literals
will be moved to the golden file `collations.sql`

**Review by cloud-fan — APPROVED** [review]

File: sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala [hunk 5]
```diff
@@ -17,135 +17,130 @@
...
+        assert(sql(query).schema.fields.head.dataType.sameType(StringType(t.collation)))
+      }
     })
-    // Collation mismatch
-    checkError(
-      exception = intercept[AnalysisException] {
-        sql("SELECT concat_ws(' ', collate('Spark', 'UTF8_LCASE'), collate('SQL', 'UNICODE'))")
-      },
-      errorClass = "COLLATION_MISMATCH.EXPLICIT",
-      parameters = Map("explicitTypes" -> "`string collate UTF8_LCASE`, `string collate UNICODE`")
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala:63]: Did you move this test somewhere or it is duplicate of another test?

File: sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala [hunk 6]
```diff
@@ -17,135 +17,130 @@
...
       ConcatWsTestCase(" ", Array("Spark", "SQL"), "UNICODE", "Spark SQL"),
       ConcatWsTestCase(" ", Array("Spark", "SQL"), "UNICODE_CI", "Spark SQL")
     )
     testCases.foreach(t => {
-      val arrCollated = t.a.map(s => s"collate('$s', '${t.c}')").mkString(", ")
-      var query = s"SELECT concat_ws(collate('${t.s}', '${t.c}'), $arrCollated)"
-      // Result & data type
-      checkAnswer(sql(query), Row(t.result))
-      assert(sql(query).schema.fields.head.dataType.sameType(StringType(t.c)))
-      // Implicit casting
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala:48]: Did you remove the implicit casting tests? If so, please, explain why.

File: sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala [hunk 7]
```diff
@@ -17,135 +17,130 @@
...
+        case s: String => Literal.create(s, StringType(t.collation))
+      }
+      val sepExpr = Literal.create(t.sep, StringType(t.collation))
+      checkEvaluation(ConcatWs(sepExpr +: inputExprs.toIndexedSeq), t.result)
+      // E2E SQL test.
+      withSQLConf(SQLConf.DEFAULT_COLLATION.key -> t.collation) {
+        val array = t.arrayStr.map(s => s"'$s'").mkString(", ")
+        val query = s"select concat_ws('${t.sep}', $array)"
+        checkAnswer(sql(query), Row(t.result))
+        assert(sql(query).schema.fields.head.dataType.sameType(StringType(t.collation)))
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala:57]: Actually you repeats this 2 lines many times, how about to create a private helper method.

File: sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala (hunk 5 above)
Comment (uros-db) [reply sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala:63]: As I mentioned in the comment above, these tests will be moved to the `collations.sql` golden file, as part of a separate PR

File: sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala (hunk 6 above)
Comment (uros-db) [reply sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala:48]: As I mentioned in the comment above, these tests will be moved to the `collations.sql` golden file, as part of a separate PR

File: sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala (hunk 7 above)
Comment (uros-db) [reply sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala:57]: It was already done like this in the current, and many other test suites, so I'm not inclined to change it

These lines are rather simple, so I think repeating 1 line instead of 2 lines across the suite is not going to help a lot, also - having to jump to a single private helper method in all tests may negatively impact debugging

**Review by MaxGekk — APPROVED** [review]

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @uros-db and @cloud-fan for review.
